src/libraries/System.Collections/src/System/Collections/Generic/PriorityQueue.cs

```diff
@@ -0,0 +1,514 @@
+        /// </summary>
+        public void EnqueueRange(IEnumerable<TElement> elements, TPriority priority)
+        {
+            if (elements is null)
+            {
+                throw new ArgumentNullException(nameof(elements));
+            }
+
+            foreach (var element in elements)
+            {
+                this.Enqueue(element, priority);
+            }
+        }
+
+        /// <summary>
+        /// Removes all items from the priority queue.
+        /// </summary>
+        public void Clear()
+        {
+            this.nodes.Clear();
+        }
+
+        /// <summary>
+        /// Ensures that the priority queue has the specified capacity
+        /// and resizes its underlying array if necessary.
+        /// </summary>
+        public void EnsureCapacity(int capacity)
+        {
+            if (capacity < 0)
+            {
+                throw new ArgumentOutOfRangeException(nameof(capacity));
+            }
+
+            if (capacity <= this.nodes.Count)
+            {
+                return;
+            }
+
+            this.nodes.Capacity = capacity;
+        }
+
+        /// <summary>
+        /// Sets the capacity to the actual number of items in the priority queue,
+        /// if that is less than 90 percent of current capacity.
+        /// </summary>
+        public void TrimExcess()
+        {
+            this.nodes.TrimExcess();
+        }
+
+        /// <summary>
+        /// Removes the node at the specified index.
+        /// </summary>
+        private void Remove(int indexOfNodeToRemove)
+        {
+            // The idea is to replace the specified node by the very last
+            // node and shorten the array by one.
+
+            var lastNodeIndex = this.GetLastNodeIndex();
+            var lastNode = this.nodes[lastNodeIndex];
+            this.nodes.RemoveAt(lastNodeIndex);
+
+            // In case we wanted to remove the node that was the last one,
+            // we are done.
+
+            if (indexOfNodeToRemove == lastNodeIndex)
+            {
+                return;
+            }
+
+            // Our last node was erased from the array and needs to be
+            // inserted again. Of course, we will overwrite the node we
+            // wanted to remove. After that operation, we will need
+            // to restore the heap property (in general).
+
+            var nodeToRemove = this.nodes[indexOfNodeToRemove];
+
+            var relation = this.Comparer.Compare(lastNode.priority, nodeToRemove.priority);
+            this.PutAt(lastNode, indexOfNodeToRemove);
+
+            if (relation < 0)
+            {
+                this.MoveUp(lastNode, indexOfNodeToRemove);
+            }
+            else
+            {
+                this.MoveDown(lastNode, indexOfNodeToRemove);
+            }
+        }
+
+        /// <summary>
+        /// Puts a node at the specified index.
+        /// </summary>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
```

Comment:
Why are all of these AggressiveInlinings necessary?

Reply:
Possibly this is an example of me trying to outsmart the compiler and in the end doing something stupid 😄 Was trying to achieve inlining this method in its usages, given that it's just a simple expression, to avoid unnecessary calls to this method. The compiler will be smart enough to inline it without the attributes?

Reply:
> The compiler will be smart enough to inline it without the attributes?

Yes

Reply:
We only apply this when (1) it is proven useful in benchmarks && (2) we have reason to believe this will realistically apply to real apps, which typically it doesn't (it is good as making microbenchmarks look good)